include/envoy/upstream/cluster_manager.h

```diff
@@ -73,6 +73,44 @@ using ClusterUpdateCallbacksHandlePtr = std::unique_ptr<ClusterUpdateCallbacksHa
 
 class ClusterManagerFactory;
 
+// These are per-cluster per-thread, so not "global" stats.
+struct ClusterConnectivityState {
+  ~ClusterConnectivityState() {
+    ASSERT(pending_streams_ == 0);
+    ASSERT(active_streams_ == 0);
+    ASSERT(connecting_capacity_ == 0);
+  }
+
+  template <class T> void checkAndDecrement(T& value, uint32_t delta) {
+    ASSERT(delta <= value);
+    value -= delta;
+  }
+
+  template <class T> void checkAndIncrement(T& value, uint32_t delta) {
+    ASSERT(std::numeric_limits<T>::max() - value > delta);
+    value += delta;
+  }
+
+  void incrPendingStreams(uint32_t delta) { checkAndIncrement<uint32_t>(pending_streams_, delta); }
+  void decrPendingStreams(uint32_t delta) { checkAndDecrement<uint32_t>(pending_streams_, delta); }
+  void incrConnectingCapacity(uint32_t delta) {
+    checkAndIncrement<uint64_t>(connecting_capacity_, delta);
+  }
+  void decrConnectingCapacity(uint32_t delta) {
+    checkAndDecrement<uint64_t>(connecting_capacity_, delta);
+  }
+  void incrActiveStreams(uint32_t delta) { checkAndIncrement<uint32_t>(active_streams_, delta); }
+  void decrActiveStreams(uint32_t delta) { checkAndDecrement<uint32_t>(active_streams_, delta); }
+
+  // Tracks the number of pending streams for this ClusterManager.
+  uint32_t pending_streams_{};
+  // Tracks the number of active streams for this ClusterManager.
+  uint32_t active_streams_{};
+  // Tracks the stream capacity if all connecting connections were connected but
+  // excluding streams which are in use.
+  uint64_t connecting_capacity_{};
```

Comment:
nit: can you name this `connecting_stream_capacity_` or `connection_stream_capacity_` and similar for the accessor methods? I was confused when I first started to look at the code below.